### ISSUE TITLE
Feature: Render children only in `itemBuilder` of GridView.builder

### ISSUE THREAD

**Issue by AmineLAHRIM** [karvulf/flutter-reorderable-grid-view#43]
even use 
`ReorderableType reorderableType = ReorderableType.gridViewBuilder;`
so it means the library is not really using the viewport of rendering UI as GridView.Builder do

please if you can solve this issue will be game-changing for Animated GridView.

**Comment by karvulf** [comment]
Hello @AmineLAHRIM 
Thank you for opening the issue. I look into it and get in touch later

**Comment by karvulf** [comment]
I found two issues. The first one is, that I am using the hashcode of every child key. This doesn't guarantee that every would be unique. But I could solve that issue by using the key instead of the hashcode.

The second problem is the behaviour, when using a `GridView.builder` in combination with `ReorderableBuilder`. Currently I expect to get all children that should be rendered inside `GirdView.Builder`. This leads to a performance problem when having a lot of children.
I will think about a solution to prevent that.
@AmineLAHRIM

**Comment by karvulf** [comment]
I have some updates about the other issue. I am looking forward to a solution that could work:

I will add a new function called `itemBuilder` which can be used for `GridView.builder`. This function will be called when `GridView.builder` is calling his own `itemBuilder`. The function returns the created widget with the functionality for drag and drop and other things.
This seems to work but there are still some things I have to investigate but currently it doesn't look that bad. If I have a working solution to test, I will come back @AmineLAHRIM

**Comment by karvulf** [comment]
My optimism is gone a bit because the part to animated added or removed items is not that easy as I thought at first. My main problem is that I don't have enough control about the rendered children when they are removed or if their position has changed.
So probably this functionality could come with the next major version but it would take some time to implement that.

I will add this as feature to my roadmap, sorry for that @AmineLAHRIM

**Comment by karvulf** [comment]
Are you only interested for the animation when adding or removing items or would it also help if only the drag and drop functionality works? Because that would be possible to implement but I am not sure if that is the thing you wished for @AmineLAHRIM

**Comment by AmineLAHRIM** [comment]
Only add and remove, and i know there too mch work their, because if not you will find flutter team did it as, andoid Team did with RecycleView

**Comment by andreBalh** [comment]
What did the flutter team? I think one word is missing :D @AmineLAHRIM

**Comment by AmineLAHRIM** [comment]
Flutter team didn't do nothing about it
But Android Team did Recycle view that fix this issue

**Comment by zezo357** [comment]
i would love to use the reorder animation with my gridview builder, from what i found this is the only package that allows gridview changes to be animated.

btw to be 100% sure, if the list changes without insert or remove its animated right?? (if they have the same key ofcourse)

so i waiting for that release ♥ hope it comes soon

**Comment by karvulf** [comment]
If the list changes, e. g. an item is replaced or changes his position, that would be animated. You could also use your current `GridView.builder` but the performance could be bad if you have a lot of items in your GridView. @zezo357

**Comment by zezo357** [comment]
> If the list changes, e. g. an item is replaced or changes his position, that would be animated. You could also use your current `GridView.builder` but the performance could be bad if you have a lot of items in your GridView. @zezo357 

I have lots of items in my gridview and animations so I can't use the package In The current state🥲🥲🥲

So I will wait for the update for better performance ❤️

**Comment by karvulf** [comment]
I have some updates for you @zezo357 and @AmineLAHRIM 
Currently I am working for the next major update and I am trying to optimize the `ReorderableGridView`. To do that I reimplemented the code.
That means that currently, the drag and drop functionality is not working but this one was never the big problem.
So I started with the animated functionality, especially to get a working solution for `GridView.builder`. I think I found one, that has still some problems with the animation but just small ones.
So if you are just interested for the animated side, you can test the current solution and give me feedback what you think about that.
There is an example app, that shows how you add `ReorderableBuilder.builder` for `GridView.builder`.
You can test the branch by adding the following to your `pubspec.yaml`:

```yaml
flutter_reorderable_grid_view:
  git:
    url: https://github.com/karvulf/flutter-reorderable-grid-view.git
    ref: 24ce256ec4a3a017768ddd7805cb660888759f42
```

**Comment by zezo357** [comment]
@karvulf Just tested it, I love it ♥ thank you very much♥

**Comment by karvulf** [comment]
Currently I think the animation part for GridViews is almost working fluently, I still need to implement a correct animation behavior considering the current scrolling position inside the GridView but the rest should work without weird bugs.
As next step, I will implement the Drag and Drop functionality for all GridViews with some more animations that are documented in other issues.
When this is done, I will continue with `Wrap` and hopefully I will find a good solution for this widget because I am pretty sure that it will be much harder to find a solution that is working in a normal behavior. @zezo357 @AmineLAHRIM

**Comment by zezo357** [comment]
For now the package does what I needed, and does so very well.

The drag and drop will be very good since it will do what the package is implemented for ❤️

**Comment by karvulf** [comment]
Hello @zezo357 

I continued on my implementation for the release `4.0.0` and have some news. The drag and drop functionality is added again in combination with the animation for all types of GridView.

**Comment by karvulf** [comment]
#60

**Comment by zezo357** [comment]
> #60 

Thank you ❤️

**Comment by zezo357** [comment]
I have testing the pre release 4 for a while now (before it was named 5)

Just saying that to avoid confusion if it was fixed 



I have built the app with it but 


when I am scrolling the scroll is not smooth I will try to take a video recording for it 



Also sometimes the changes doesn't get animated(most of the time it does )

95% does animate
5% doesn't 

Don't know why since it happened at release and didn't find any errors related to it

**Comment by zezo357** [comment]
I will test on version 5 too, but just wanted to note what I have seen Incase they can be improved



thank you for the awesome package.


If I find a problem and solve it I will add a pull request, if you don't mind it

**Comment by karvulf** [comment]
Ah interesting, soon I will release `5.0.0-dev.2`, so maybe there is a little improvement. I had to change the pre release to `5.0.0-dev.1` because of the flutter update. Maybe your version was increased to `4.0.0` and that changed the performance because this update doesn't contain any improvements for the `GridView.builder`.

**Comment by zezo357** [comment]
> Ah interesting, soon I will release `5.0.0-dev.2`, so maybe there is a little improvement. I had to change the pre release to `5.0.0-dev.1` because of the flutter update. Maybe your version was increased to `4.0.0` and that changed the performance because this update doesn't contain any improvements for the `GridView.builder`. @zezo357 

Ok will check again with `5.0.0-dev.2` when it gets released, any idea when it will be available?

**Comment by karvulf** [comment]
I will try make the update this week, but before I check your issue. Did it happen on android or iOS? @zezo357

**Comment by zezo357** [comment]
> I will try make the update this week, but before I check your issue. Did it happen on android or iOS? @zezo357 

Android

Pixel 6 to be specific so the screen is 90hz

**Comment by karvulf** [comment]
So I added some changes to the release `5.0.0-dev.2` and it also optimizes the reordering. The release is available to test.

I tried with a simple example of `GridView.builder` containing 100.000 items. I didn't have any performance issues. So possibly, it could be helpful to see how you use `ReorderableBuilder` and how you build your children.
Because currently the package should only cache the items that are loaded through `GridView.builder` with `itemBuilder`.
But what I saw, when I am using the scrolling functionality while dragging, sometimes children are not rendered correctly (they are not visible, but still there). But I don't think that you meant this issue. So thank you already for helping me to create a good working release! :) @zezo357

**Comment by karvulf** [comment]
And the most important part to get a working solution for the callback `onReorder`:
You have to change
```dart
void _handleReorder(List<OrderUpdateEntity> onReorderList) {
  for (final reorder in onReorderList) {
    final child = children.removeAt(reorder.oldIndex);
    children.insert(reorder.newIndex, child);
  }
  setState(() {});
}
``` 
to
```dart
void _handleReorder(ReorderedListFunction reorderedListFunction) {
  setState(() {
    children = reorderedListFunction(children) as List<int>;
  });
}
```
Make sure to cast the list to the type you use for your list.

**Comment by zezo357** [comment]
> And the most important part that your code has no errors after upgrading to `5.0.0-dev.2` is to change the callback `onReorder`:
> You have to change
> ```dart
> void _handleReorder(List<OrderUpdateEntity> onReorderList) {
>   for (final reorder in onReorderList) {
>     final child = children.removeAt(reorder.oldIndex);
>     children.insert(reorder.newIndex, child);
>   }
>   setState(() {});
> }
> ``` 
> to
> ```dart
> void _handleReorder(ReorderedListFunction reorderedListFunction) {
>   setState(() {
>     children = reorderedListFunction(children) as List<int>;
>   });
> }
> ```
> Make sure to cast the list to the type you use for your list.

I am not using reorder, I am only using the part of animating grid changes 

Should I implement this too? 

Also is the example updated? Maybe I can check if there is something not right with my code 


Sadly I can't check today, but thank you for the update will test it as soon as possible ❤️

**Comment by zezo357** [comment]
> So I added some changes to the release `5.0.0-dev.2` and it also optimizes the reordering. The release is available to test.
> 
> I tried with a simple example of `GridView.builder` containing 100.000 items. I didn't have any performance issues. So possibly, it could be helpful to see how you use `ReorderableBuilder` and how you build your children.
> Because currently the package should only cache the items that are loaded through `GridView.builder` with `itemBuilder`.
> But what I saw, when I am using the scrolling functionality while dragging, sometimes children are not rendered correctly (they are not visible, but still there). But I don't think that you meant this issue. So thank you already for helping me to create a good working release! :) @zezo357 

Also I faced the problem of the items not showing,but being there.

Didn't know why it happens and its not frequent.

If I saw an error related to it I will let you know

**Comment by karvulf** [comment]
Ahhhh totally thought you are using the drag and drop functionality. Then don't worry, you don't have to use it. 
It would be still interesting why you have the performance issue. I will check again this with my example code just by scrolling through the items. @zezo357

**Comment by zezo357** [comment]
> Ahhhh totally thought you are using the drag and drop functionality. Then don't worry, you don't have to use it. 
> It would be still interesting why you have the performance issue. I will check again this with my example code just by scrolling through the items. @zezo357 

It's smooth, but sometimes the scrolling lags, so I am trying to figure if it's a problem by me or you 

So i will test again with the new release and I hope its fixed

**Comment by karvulf** [comment]
Then I think this issue will be still there. So if you could show me how your code looks, especially for the children part because maybe you are having children with images and a lot of other stuff, this could also be a problem. But I will think about the logic, maybe I can figure something out, but that will happen in the next weeks, because I am still busy :'D @zezo357

**Comment by zezo357** [comment]
> Then I think this issue will be still there. So if you could show me how your code looks, especially for the children part because maybe you are having children with images and a lot of other stuff, this could also be a problem. But I will think about the logic, maybe I can figure something out, but that will happen in the next weeks, because I am still busy :'D @zezo357 

The children are a stack of widgets and image yes.

I will attach a screenshot of my ui 

Good luck with your stuff too.

**Comment by zezo357** [comment]
![Screenshot_20230207-232710.png](https://user-images.githubusercontent.com/25157308/217369514-8edc98d0-9a03-4f11-b6eb-546b4b5a0489.png)

This is my ui

**Comment by zezo357** [comment]
I just tested the disappearing elements didn't happen again.

For the scrolling problem I am thinking it may be related to me enabling impler so I will test again without it to be 100% sure 

But I noticed an error in my logger 
```
[ERROR] | 11:51:26 424ms | Unhandled Exception
Null check operator used on a null value
StackTrace: #0      AnimationController.stop (package:flutter/src/animation/animation_controller.dart:785)
#1      AnimationController.value= (package:flutter/src/animation/animation_controller.dart:367)
#2      AnimationController.reset (package:flutter/src/animation/animation_controller.dart:388)
#3      _ReorderableAnimatedPositionedState._compareUpdatedReorderableEntity.<anonymous closure> (package:flutter_reorderable_grid_view/widgets/reorderable_animated_positioned.dart:131)
#4      SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1289)
#5      SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1227)
#6      SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1076)
#7      _rootRun (dart:async/zone.dart:1398)
#8      _CustomZone.run (dart:async/zone.dart:1300)
#9      _CustomZone.runGuarded (dart:async/zone.dart:1208)
#10     _invoke (dart:ui/hooks.dart:147)
#11     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:338)
#12     _drawFrame (dart:ui/hooks.dart:112)
}
```

**Comment by karvulf** [comment]
Ok I will investigate that, I have some ideas @zezo357

**Comment by zezo357** [comment]
> Ok I will investigate that, I have some ideas @zezo357 

Thank you very much ❤️

**Comment by zezo357** [comment]
> Ok I will investigate that, I have some ideas @zezo357 

Thank you very much ❤️

**Comment by abdelaziz-mahdy** [comment]
currently i see that when i scroll all of the list items gets re-rendered searching through the code i found this https://github.com/karvulf/flutter-reorderable-grid-view/blob/6fe55426c848b1e54f40d6ec973d02b60e7ffdff/lib/widgets/reorderable_builder.dart#L494
is the setstate call, do you think rendering can be reduced to only items that finished, instead of all of the widget, i think that is the cause of the performance issue i mentioned before

**Comment by karvulf** [comment]
Hello @abdelaziz-mahdy 
Thanks for mentioning that, I will check it in the weekend :)

**Comment by karvulf** [comment]
I checked what you mentioned @abdelaziz-mahdy 
and I agree with you that this rebuilt for all the items can cause issues.
I will think about a better solution to improve the rendering so that only the widget rebuilds itself which was updated.

**Comment by abdelaziz-mahdy** [comment]
> I checked what you mentioned @abdelaziz-mahdy and I agree with you that this rebuilt for all the items can cause issues. I will think about a better solution to improve the rendering so that only the widget rebuilds itself which was updated.

glad i was able to find the cause, i tried to make the element itself setstate but it caused issues, and since i dont full understand the logic of the widget tree i couldnt fix it, will wait for your fix and would gladly test it <3

**Comment by karvulf** [comment]
But I cannot promise that it will fix the performance issue 😄 @abdelaziz-mahdy

**Comment by abdelaziz-mahdy** [comment]
> But I cannot promise that it will fix the performance issue 😄 @abdelaziz-mahdy 

Understandable, but from my testing that was the only re render that may cause it 😅

**Comment by karvulf** [comment]
I just released version `5.0.0` and will create a new issue related to this one @abdelaziz-mahdy
